pkg/controller/stacks/stack/stack.go

```diff
@@ -160,29 +188,169 @@ func (h *stackHandler) update(ctx context.Context) (reconcile.Result, error) {
+
+		rules := []rbacv1.PolicyRule{}
+		for _, crd := range h.ext.Spec.CRDs {
+			rules = append(rules, rbacv1.PolicyRule{
+				APIGroups: []string{crd.GroupVersionKind().Group},
+				Resources: []string{crd.Kind},
+				Verbs:     roleVerbs[persona],
+			})
+		}
+		name := fmtPersonaRole(h.ext, persona)
+
+		var crossplaneScope string
+
+		if h.isNamespaced() {
+			crossplaneScope = "namespace"
+		} else {
+			crossplaneScope = "environment"
+		}
+
+		aggregationLabel := fmt.Sprintf(labelAggregateFmt, crossplaneScope, persona)
+
+		cr := &rbacv1.ClusterRole{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:   name,
+				Labels: labelsCopy,
+			},
+			Rules: rules,
+		}
+
+		meta.AddLabels(cr, map[string]string{
+			aggregationLabel: "true",
+		})
+
+		meta.AddLabels(cr, stacks.ParentLabels(h.ext))
+
+		if h.isNamespaced() {
+			meta.AddLabels(cr, map[string]string{
+				fmt.Sprintf(labelNamespaceFmt, h.ext.GetNamespace()): "true",
+			})
+		}
+
+		if err := h.kube.Create(ctx, cr); err != nil && !kerrors.IsAlreadyExists(err) {
+			return errors.Wrap(err, "failed to create persona cluster roles")
+		}
+	}
+	return nil
+}
+
+func generateNamespaceClusterRoles(stack *v1alpha1.Stack) (roles []*rbacv1.ClusterRole) {
+	personas := []string{"admin", "edit", "view"}
+
+	namespaced := (stack.Spec.PermissionScope == string(apiextensions.NamespaceScoped))
+	if !namespaced {
+		return
+	}
+
+	nsName := stack.GetNamespace()
+
+	for _, persona := range personas {
+		name := fmt.Sprintf("crossplane:ns:%s:%s", nsName, persona)
+
+		labels := map[string]string{
+			fmt.Sprintf(labelNamespaceFmt, nsName): "true",
+			labelScope:                             "namespace",
+		}
+
+		if persona == "admin" {
+			labels[fmt.Sprintf(labelAggregateFmt, "crossplane", persona)] = "true"
+		}
+
+		role := &rbacv1.ClusterRole{
+			AggregationRule: &rbacv1.AggregationRule{
+				ClusterRoleSelectors: []metav1.LabelSelector{
+					{
+						MatchLabels: map[string]string{
+							fmt.Sprintf(labelAggregateFmt, "namespace", persona): "true",
+							fmt.Sprintf(labelNamespaceFmt, nsName):               "true",
+						},
+					},
+					{
+						MatchLabels: map[string]string{
+							fmt.Sprintf(labelAggregateFmt, "namespace-default", persona): "true",
+						},
+					},
+				},
+			},
+
+			// TODO(displague) set parent labels?
+			ObjectMeta: metav1.ObjectMeta{
+				Name:   name,
+				Labels: labels,
+			},
+		}
+
+		roles = append(roles, role)
+	}
+
+	return roles
+}
+
+func (h *stackHandler) createNamespaceClusterRoles(ctx context.Context) error {
+	ns := &corev1.Namespace{}
+	nsName := h.ext.GetNamespace()
+
+	if err := h.kube.Get(ctx, types.NamespacedName{Name: nsName}, ns); err != nil {
+		return errors.Wrapf(err, "failed to get namespace %q for stackinstall %q", nsName, h.ext.GetName())
+	}
+	roles := generateNamespaceClusterRoles(h.ext)
+
+	for _, role := range roles {
+		role.SetOwnerReferences([]metav1.OwnerReference{
+			{
+				APIVersion: "v1",
+				Kind:       "Namespace",
```

Comment:
is setting the ownerRef to the namespace explicitly required? I thought it was implicit that when a namespace is deleted that all resources within the namespace are deleted too.

Reply:
We are creating a Cluster scoped `ClusterRole` with a Cluster scoped `Namespace` as the owner.

This is not the same as setting a namespaced `Role` within a `Namespace`, where the behavior you suggested would apply.

If the namespace is deleted, the clusterroles that are bound to that namespace (by name only) should go away, which is why OwnerReferences are a convenient approach.

When we take on https://github.com/crossplaneio/crossplane/issues/1124 we could consider letting the stack-manager clean up these clusterroles if it detects that the namespace was deleted, but I think we will still find OwnerReferences to be more reliable (especially if stack-manager is uninstalled).